python/pyspark/testing/utils.py

```diff
@@ -464,23 +467,42 @@ def assertDataFrameEqual(
         raise PySparkAssertionError(
             error_class="INVALID_TYPE_DF_EQUALITY_ARG",
             message_parameters={
-                "expected_type": Union[DataFrame, ps.DataFrame, List[Row]],
+                "expected_type": "Union[DataFrame, ps.DataFrame, List[Row]]",
                 "arg_name": "expected",
                 "actual_type": None,
             },
         )
 
+    has_pandas = False
     try:
-        # If Spark Connect dependencies are available, allow Spark Connect DataFrame
-        from pyspark.sql.connect.dataframe import DataFrame as ConnectDataFrame
+        # If pandas dependencies are available, allow pandas or pandas-on-Spark DataFrame
+        import pyspark.pandas as ps
+        import pandas as pd
+        from pyspark.testing.pandasutils import PandasOnSparkTestUtils
 
-        if isinstance(actual, ps.DataFrame) or isinstance(expected, ps.DataFrame):
+        has_pandas = True
+    except Exception:
+        # no pandas, so we won't call pandasutils functions
+        pass
+
+    if has_pandas:
+        if (
+            isinstance(actual, pd.DataFrame)
+            or isinstance(expected, pd.DataFrame)
+            or isinstance(actual, ps.DataFrame)
+            or isinstance(expected, ps.DataFrame)
+        ):
             # handle pandas DataFrames
             # assert approximate equality for float data
-            return assertPandasOnSparkEqual(
-                actual, expected, checkExact=False, checkRowOrder=checkRowOrder
+            return PandasOnSparkTestUtils().assert_eq(
+                actual, expected, almost=True, rtol=rtol, atol=atol, check_row_order=checkRowOrder
             )
-        elif not isinstance(actual, (DataFrame, ConnectDataFrame, list)):
+
+    try:
+        # If Spark Connect dependencies are available, allow Spark Connect DataFrame
+        from pyspark.sql.connect.dataframe import DataFrame as ConnectDataFrame
+
+        if not isinstance(actual, (DataFrame, ConnectDataFrame, list)):
```

Comment:
I think we should reassign `DataFrame` to `ConnectDataFrame` if `is_remote` is `True`. cc @itholic

Reply:
Yeah, we have `pyspark.sql.utils.get_dataframe_class` to get proper DataFrame class.

e.g.
```python
from pyspark.sql.utils import get_dataframe_class
...
        SparkDataFrame = get_dataframe_class()
        if not isinstance(actual, (SparkDataFrame, list)):
...
```

Reply:
Do you think it would be useful for users to migrate to use Spark Connect? I.e checking the dataframe results are the same.

Reply:
sounds good, i'll allow for spark DataFrame or connect DataFrame if is_remote() is True, to allow for these comparisons